### PR TITLE
ipmitool: enterprise-numbers: Add enterprise code for IPS

### PR DESCRIPTION
**Pull request by lxwinspur**
The enterprise code of IPS is in the file
http://www.iana.org/assignments/enterprise-numbers/enterprise-numbers And needs to be synchronized to OpenBMC.

Signed-off-by: George Liu <liuxiwei@inspur.com>

### CODE REVIEW

**Comment by geissonator** [comment]
Link to upstream commit as well George? Please make the PR's against 1020. Reed will make sure it gets into the correct service pack (and 1030 and beyond).

But, the problem with this change is it's global to all p10bmc machines, which means even non-IPS p10bmc machines would return this. As we want to keep our p10bmc flash image common, and this is a compile-time option, it's not clear to me how we can support different IANA numbers in a single image.

This change also highlights that we never did a IANA for p10bmc machines! Oh well, I'm just not sure how critical this is. If it is critical then we'll need to either create a new flash image for IPS, or find a way to make this runtime configurable within IPMI.

**Comment by lxwinspur** [comment]
> Link to upstream commit as well George? 

Yes, Patrick has pushed a new patch for it, and then I will rebase this patch and add the IPS enterprise code.

> Please make the PR's against 1020. Reed will make sure it gets into the correct service pack (and 1030 and beyond).

Sorry, you mean that should I push a PR to 1020 branch instead of 1020.10 branch right?

> 
> But, the problem with this change is it's global to all p10bmc machines, which means even non-IPS p10bmc machines would return this. As we want to keep our p10bmc flash image common, and this is a compile-time option, it's not clear to me how we can support different IANA numbers in a single image.

Yes, this is what bothers us too, does IBM have a good way to follow IANA for IPS compatibility?

> 
> This change also highlights that we never did a IANA for p10bmc machines! Oh well, I'm just not sure how critical this is. If it is critical then we'll need to either create a new flash image for IPS, or find a way to make this runtime configurable within IPMI.

Sure, Thanks @geissonator

**Comment by rfrandse** [comment]
@lxwinspur for this change 1020.10 branch works.

**Comment by lxwinspur** [comment]
@geissonator @rfrandse upstream has been merged: https://gerrit.openbmc.org/c/openbmc/openbmc/+/57071

**Comment by geissonator** [comment]
> Yes, this is what bothers us too, does IBM have a good way to follow IANA for IPS compatibility?

No, you'll need to work with the IPMI maintainers to find a way to support dynamic IANA. Probably be best if we could store it in a inventory item or d-bus setting that we could override when we see it's an IPS system.

> @geissonator @rfrandse upstream has been merged: https://gerrit.openbmc.org/c/openbmc/openbmc/+/57071

You didn't get the dev_id.json into upstream which is the issue we need to resolve with this PR. 

Also, @sbteeks found when he was debugging the missing IBM support that there are actually 2 places IANA come from. This dev_id.json seems to provide the raw ipmitool command version, but somewhere else, there is what comes back from "ipmitool mc info".

**Review by geissonator — CHANGES_REQUESTED** [review]
Need to resolve open questions before we can merge this. If you want to just get the enterprise-numbers change in, please do a separate PR for just that.

**Comment by lxwinspur** [comment]
> Need to resolve open questions before we can merge this. If you want to just get the enterprise-numbers change in, please do a separate PR for just that.

Pushed 2 commits, please review it, thanks ;)

**Comment by lxwinspur** [comment]
@mzipse 
Please check this issue, thanks

**Comment by sbteeks** [comment]
> > Yes, this is what bothers us too, does IBM have a good way to follow IANA for IPS compatibility?
> 
> No, you'll need to work with the IPMI maintainers to find a way to support dynamic IANA. Probably be best if we could store it in a inventory item or d-bus setting that we could override when we see it's an IPS system.
> 
> > @geissonator @rfrandse upstream has been merged: https://gerrit.openbmc.org/c/openbmc/openbmc/+/57071
> 
> You didn't get the dev_id.json into upstream which is the issue we need to resolve with this PR.
> 
> Also, @sbteeks found when he was debugging the missing IBM support that there are actually 2 places IANA come from. This dev_id.json seems to provide the raw ipmitool command version, but somewhere else, there is what comes back from "ipmitool mc info".

While I don't know all details/discussions on this, but if we end up with dev_id_ips.json  does 'mc info' matches with IPS IANA number? If not, are we ok to get two different IANA numbers depending on which NetFn/cmd used?

File: meta-ibm/recipes-phosphor/ipmi/phosphor-ipmi-config/p10bmc/changeIana.sh
Comment (sbteeks) [inline meta-ibm/recipes-phosphor/ipmi/phosphor-ipmi-config/p10bmc/changeIana.sh:12]: Do we get different IANA numbers if 'model != "J0"'?

All I can see is echo different strings when model is S0/D0 and in the else part...

If not, we can just check something like:
if [ "${model[1]}" = '"J0"' ]; then
    echo "loading IPS IANA"
    mount --bind ${DIR}/dev_id_ips.json ${DIR}/dev_id.json
fi

**Comment by lxwinspur** [comment]
> > Also, @sbteeks found when he was debugging the missing IBM support that there are actually 2 places IANA come from. This dev_id.json seems to provide the raw ipmitool command version, but somewhere else, there is what comes back from "ipmitool mc info".
> 
> While I don't know all the details/discussions on this, but if we end up with dev_id_ips.json does 'mc info' matches with IPS IANA number? If not, are we ok to get two different IANA numbers depending on which NetFn/cmd used?

Nope @sbteeks , There are 2 dev_id.json and the dev_id.json belongs to IBM and dev_id_ips.json belongs to IPS.
We will check whether the current system is IBM or IPS via the property value of VPD, If it is IBM, load dev_id.json, If it is IPS, load dev_id_ips.json. no influence between them.

File: meta-ibm/recipes-phosphor/ipmi/phosphor-ipmi-config/p10bmc/changeIana.sh
Comment (lxwinspur) [reply meta-ibm/recipes-phosphor/ipmi/phosphor-ipmi-config/p10bmc/changeIana.sh:12]: > Do we get different IANA numbers if 'model != "J0"'?

I think Rainier will support OEM Systems as well, like this:
https://github.com/ibm-openbmc/openbmc/blob/1020.20/meta-ibm/recipes-phosphor/webui/webui-vue/changeLogo.sh

> 
> All I can see is echo different strings when model is S0/D0 and in the else part...
> 
> If not, we can just check something like: if [ "${model[1]}" = '"J0"' ]; then echo "loading IPS IANA" mount --bind DIR/devidips.json{DIR}/dev_id.json fi

@geissonator @mzipse 
What do you think?

**Comment by sbteeks** [comment]
> > > Also, @sbteeks found when he was debugging the missing IBM support that there are actually 2 places IANA come from. This dev_id.json seems to provide the raw ipmitool command version, but somewhere else, there is what comes back from "ipmitool mc info".
> > 
> > 
> > While I don't know all the details/discussions on this, but if we end up with dev_id_ips.json does 'mc info' matches with IPS IANA number? If not, are we ok to get two different IANA numbers depending on which NetFn/cmd used?
> 
> Nope @sbteeks , There are 2 dev_id.json and the dev_id.json belongs to IBM and dev_id_ips.json belongs to IPS. We will check whether the current system is IBM or IPS via the property value of VPD, If it is IBM, load dev_id.json, If it is IPS, load dev_id_ips.json. no influence between them.

I got that part, but concern is about 'mc info' not matching IPS IANA number when the model is "J0". Not clear to me whether they both match or not. You should test with the following commands just to make sure IANA numbers match:
"ipmitool -I lanplus -C 17 -p 623 -U user -P password -H host raw 0x06 0x01"
"ipmitool -I lanplus -C 17 -p 623 -U user -P password -H host mc info"

**Comment by lxwinspur** [comment]
> > > > Also, @sbteeks found when he was debugging the missing IBM support that there are actually 2 places IANA come from. This dev_id.json seems to provide the raw ipmitool command version, but somewhere else, there is what comes back from "ipmitool mc info".
> > > 
> > > 
> > > While I don't know all the details/discussions on this, but if we end up with dev_id_ips.json does 'mc info' matches with IPS IANA number? If not, are we ok to get two different IANA numbers depending on which NetFn/cmd used?
> > 
> > 
> > Nope @sbteeks , There are 2 dev_id.json and the dev_id.json belongs to IBM and dev_id_ips.json belongs to IPS. We will check whether the current system is IBM or IPS via the property value of VPD, If it is IBM, load dev_id.json, If it is IPS, load dev_id_ips.json. no influence between them.
> 
> I got that part, but concern is about 'mc info' not matching IPS IANA number when the model is "J0". Not clear to me whether they both match or not. You should test with the following commands just to make sure IANA numbers match: "ipmitool -I lanplus -C 17 -p 623 -U user -P password -H host raw 0x06 0x01" "ipmitool -I lanplus -C 17 -p 623 -U user -P password -H host mc info"

I don't think you should use the remote ipmitool, but the ipmitool compiled with the Rainier image, because we just added this patch when building ipmitool
https://github.com/ibm-openbmc/openbmc/pull/234/files#diff-a007994f4369179beb6d03e30735d18fd76013e5d7fc03eaabd886cd3d6c1d0fR114-R117

So, I think you should test it on Rainier image, like this;
`ipmitool mc info`

But, AFAIK, Rainier image does not compile ipmitool into it, so I think you can test it like this:
1. please check if the /usr/share/ipmi-providers/dev_id.json file is the same as the ips_dev_id.json file
2. please add ipmitool in the meta-ibm/recipes-phosphor/packagegroup
3. please test `mc info` using the following command on the rainier image:
`ipmitool mc info`

**Comment by sbteeks** [comment]
> > > > > Also, @sbteeks found when he was debugging the missing IBM support that there are actually 2 places IANA come from. This dev_id.json seems to provide the raw ipmitool command version, but somewhere else, there is what comes back from "ipmitool mc info".
> > > > 
> > > > 
> > > > While I don't know all the details/discussions on this, but if we end up with dev_id_ips.json does 'mc info' matches with IPS IANA number? If not, are we ok to get two different IANA numbers depending on which NetFn/cmd used?
> > > 
> > > 
> > > Nope @sbteeks , There are 2 dev_id.json and the dev_id.json belongs to IBM and dev_id_ips.json belongs to IPS. We will check whether the current system is IBM or IPS via the property value of VPD, If it is IBM, load dev_id.json, If it is IPS, load dev_id_ips.json. no influence between them.
> > 
> > 
> > I got that part, but concern is about 'mc info' not matching IPS IANA number when the model is "J0". Not clear to me whether they both match or not. You should test with the following commands just to make sure IANA numbers match: "ipmitool -I lanplus -C 17 -p 623 -U user -P password -H host raw 0x06 0x01" "ipmitool -I lanplus -C 17 -p 623 -U user -P password -H host mc info"
> 
> I don't think you should use the remote ipmitool, but the ipmitool compiled with the Rainier image, because we just added this patch when building ipmitool https://github.com/ibm-openbmc/openbmc/pull/234/files#diff-a007994f4369179beb6d03e30735d18fd76013e5d7fc03eaabd886cd3d6c1d0fR114-R117
> 
> So, I think you should test it on Rainier image, like this; `ipmitool mc info`
> 
> But, AFAIK, Rainier image does not compile ipmitool into it, so I think you can test it like this:
> 
> 1. please check if the /usr/share/ipmi-providers/dev_id.json file is the same as the ips_dev_id.json file
> 2. please add ipmitool in the meta-ibm/recipes-phosphor/packagegroup
> 3. please test `mc info` using the following command on the rainier image:
>    `ipmitool mc info`

I'm sure you must have done the testing. Please post the output of the following:
ipmitool mc info
ipmitool raw  0x06 0x01

**Comment by lxwinspur** [comment]
> > > > > > Also, @sbteeks found when he was debugging the missing IBM support that there are actually 2 places IANA come from. This dev_id.json seems to provide the raw ipmitool command version, but somewhere else, there is what comes back from "ipmitool mc info".
> > > > > 
> > > > > 
> > > > > While I don't know all the details/discussions on this, but if we end up with dev_id_ips.json does 'mc info' matches with IPS IANA number? If not, are we ok to get two different IANA numbers depending on which NetFn/cmd used?
> > > > 
> > > > 
> > > > Nope @sbteeks , There are 2 dev_id.json and the dev_id.json belongs to IBM and dev_id_ips.json belongs to IPS. We will check whether the current system is IBM or IPS via the property value of VPD, If it is IBM, load dev_id.json, If it is IPS, load dev_id_ips.json. no influence between them.
> > > 
> > > 
> > > I got that part, but concern is about 'mc info' not matching IPS IANA number when the model is "J0". Not clear to me whether they both match or not. You should test with the following commands just to make sure IANA numbers match: "ipmitool -I lanplus -C 17 -p 623 -U user -P password -H host raw 0x06 0x01" "ipmitool -I lanplus -C 17 -p 623 -U user -P password -H host mc info"
> > 
> > 
> > I don't think you should use the remote ipmitool, but the ipmitool compiled with the Rainier image, because we just added this patch when building ipmitool https://github.com/ibm-openbmc/openbmc/pull/234/files#diff-a007994f4369179beb6d03e30735d18fd76013e5d7fc03eaabd886cd3d6c1d0fR114-R117
> > So, I think you should test it on Rainier image, like this; `ipmitool mc info`
> > But, AFAIK, Rainier image does not compile ipmitool into it, so I think you can test it like this:
> > 
> > 1. please check if the /usr/share/ipmi-providers/dev_id.json file is the same as the ips_dev_id.json file
> > 2. please add ipmitool in the meta-ibm/recipes-phosphor/packagegroup
> > 3. please test `mc info` using the following command on the rainier image:
> >    `ipmitool mc info`
> 
> I'm sure you must have done the testing. Please post the output of the following: ipmitool mc info ipmitool raw 0x06 0x01

Note: SubModel should switch to J0.
Below is my test result:
```
root@p10bmc:~# vpd-tool -r -R VSYS -K BR -O /system/chassis/motherboard
{
    "/system/chassis/motherboard": {
        "BR": "J0"
    }
}
root@p10bmc:~# cat /usr/share/ipmi-providers/dev_id.json 
{
    "id": 0,
    "revision": 128,
    "addn_dev_support": 141,
    "manuf_id": 52893,
    "prod_id": 1,
    "aux": 0
}
root@p10bmc:~# ipmitool mc info
Device ID                 : 0
Device Revision           : 0
Firmware Revision         : 0.00
IPMI Version              : 2.0
Manufacturer ID           : 52893
Manufacturer Name         : Inspur Power Systems Co.,Ltd.
Product ID                : 1 (0x0001)
Product Name              : Unknown (0x01)
Device Available          : yes
Provides Device SDRs      : yes
Additional Device Support :
    Sensor Device
    SEL Device
    FRU Inventory Device
    Chassis Device
Aux Firmware Rev Info     : 
    0x00
    0x00
    0x00
    0x00
root@p10bmc:~# ipmitool raw 0x06 0x01
 00 80 00 00 02 8d 9d ce 00 01 00 00 00 00 00
```

**Comment by sbteeks** [comment]
> root@p10bmc:~# cat /usr/share/ipmi-providers/dev_id.json 
> {
>     "id": 0,
>     "revision": 128,
>     "addn_dev_support": 141,
>     "manuf_id": 52893,
>     "prod_id": 1,
>     "aux": 0
> }
> root@p10bmc:~# ipmitool mc info
> Device ID                 : 0
> Device Revision           : 0
> Firmware Revision         : 0.00
> IPMI Version              : 2.0
> Manufacturer ID           : 52893
> Manufacturer Name         : Inspur Power Systems Co.,Ltd.
> Product ID                : 1 (0x0001)
> Product Name              : Unknown (0x01)
> Device Available          : yes
> Provides Device SDRs      : yes
> Additional Device Support :
>     Sensor Device
>     SEL Device
>     FRU Inventory Device
>     Chassis Device
> Aux Firmware Rev Info     : 
>     0x00
>     0x00
>     0x00
>     0x00
> root@p10bmc:~# ipmitool raw 0x06 0x01
>  00 80 00 00 02 8d 9d ce 00 01 00 00 00 00 00
> ```

Looks good! IANA numbers match for both "mc info" and raw response...

File: meta-ibm/recipes-phosphor/ipmi/phosphor-ipmi-config/p10bmc/changeIana.sh
Comment (geissonator) [reply meta-ibm/recipes-phosphor/ipmi/phosphor-ipmi-config/p10bmc/changeIana.sh:12]: I believe our new plan is to create a separate p10bmc-ips machine in bitbake. That will allow us to build a flash image for just IPS. This should simplify a lot of these patches as we can now just install the correct files into the bitbake image and stop keying off of vpd BR.

Comment (geissonator) [reply meta-ibm/recipes-phosphor/ipmi/phosphor-ipmi-config/p10bmc/changeIana.sh:12]: Lets just move this PR over to the IPS branch since it's done and tested.

**Comment by rfrandse** [comment]
https://github.com/ibm-openbmc/openbmc/commits/1030.00.ips
https://github.com/ibm-openbmc/openbmc/commit/f4bb4e8e8de804ae4234ac4ef227f32119a3ba43
https://github.com/ibm-openbmc/openbmc/commit/ba8eef9a3c42c5801d5d930f4f1798ee00ecc601

**Pull request closed without merging**
